apps/activejobs/app/helpers/jobs_helper.rb

```diff
@@ -4,17 +4,56 @@ def build_ganglia_link( host, start_seconds, report_type, node_num, size )
     ganglia_uri = ""
     c = OODClusters[host]
     if c && c.custom_allow?(:ganglia)
-      ganglia_cluster = OodCluster::Servers::Ganglia.new(c.custom_config(:ganglia))
-      ganglia_base = ganglia_cluster.uri.to_s
-      ganglia_server = ganglia_cluster.opt_query.fetch(:h, '') % {h: node_num }
-      ganglia_uri = ("#{ganglia_base}&z=#{size}&cs=#{start_seconds.to_s}&g=#{report_type}&h=#{ganglia_server}").html_safe
+      server = c.custom_config(:ganglia)
+      ganglia_base = "#{server[:scheme]}#{server[:host]}/#{server[:segments].join('')}"
+      cluster_query = server.fetch(:req_query, {}).map { |k,v| "#{k}=#{v}" }[0]
+      ganglia_server = server.fetch(:opt_query, {}).fetch('h', '') % {h: node_num }
+      ganglia_uri = ("#{ganglia_base}?#{cluster_query}&z=#{size}&cs=#{start_seconds.to_s}&g=#{report_type}&h=#{ganglia_server}").html_safe
     end
     ganglia_uri
   rescue
     nil
   end
 
+  def build_grafana_link(cluster, start_seconds, report_type, node_num, jobid = nil) 
+    c = OODClusters[cluster]
+    if c && c.custom_allow?(:grafana)
+      server = c.custom_config(:grafana)
+      host = node_num.split('.')[0]
+      dashboard = server[:dashboard]
+      dashboard_url = "#{dashboard['uid']}/#{dashboard['name']}"
+      query_params = {
+        orgId: server[:orgId],
+        from: "#{start_seconds}000",
+        to: 'now',
+      }
+      if ['cpu','memory'].include?(report_type)
+        url_base = 'd-solo'
+        panel_id = dashboard['panels'][report_type]
+        query_params[:panelId] = panel_id
+      else
+        url_base = 'd'
+      end
+      query_params["var-#{server[:labels]['cluster']}"] = cluster
+      query_params["var-#{server[:labels]['host']}"] = host
+      if jobid
+        jobid = jobid.split('.')[0]
```

Comment:
I have the same problem with XDMoD integration: in our case, our job ids contain extra information besides the id number that need to be dropped when using it. I'm afraid this type of transformation might not be portable between centers. But the primary goal is to support Grafana at OSC to get it out of the door, so we could revisit this in the future, and I think that the right solution, instead of custom config abstraction for these URLs, would be to have more control over the HTML template itself.

Reply:
I am fairly certain that with SLURM a job ID is always a number, and Torque it's configurable. I don't know about other schedulers we support but I would image this is rather portable.